reconcile/ocm_labels/integration.py

```diff
@@ -0,0 +1,392 @@
+from reconcile.gql_definitions.ocm_labels.organizations import (
+    query as organization_query,
+)
+from reconcile.ocm_labels.label_sources import (
+    ClusterRef,
+    LabelOwnerRef,
+    LabelSource,
+    OrgRef,
+)
+from reconcile.utils import gql
+from reconcile.utils.differ import diff_mappings
+from reconcile.utils.disabled_integrations import integration_is_enabled
+from reconcile.utils.helpers import flatten
+from reconcile.utils.ocm.clusters import discover_clusters_for_organizations
+from reconcile.utils.ocm.labels import (
+    add_label,
+    build_organization_labels_href,
+    delete_label,
+    get_org_labels,
+    update_label,
+)
+from reconcile.utils.ocm.search_filters import Filter
+from reconcile.utils.ocm_base_client import (
+    OCMAPIClientConfigurationProtocol,
+    OCMBaseClient,
+    init_ocm_base_client,
+)
+from reconcile.utils.runtime.integration import (
+    PydanticRunParams,
+    QontractReconcileIntegration,
+)
+from reconcile.utils.secret_reader import SecretReaderBase
+
+QONTRACT_INTEGRATION = "ocm-subscription-labels"
+
+
+class OcmLabelsIntegrationParams(PydanticRunParams):
+    managed_label_prefixes: list[str] = []
+
+    @validator("managed_label_prefixes")
+    def must_end_with_dot(  # pylint: disable=no-self-argument
+        cls, v: list[str]
+    ) -> list[str]:
+        return [prefix + "." if not prefix.endswith(".") else prefix for prefix in v]
+
+
+class ManagedLabelConflictError(Exception):
+    pass
+
+
+class OcmLabelsIntegration(QontractReconcileIntegration[OcmLabelsIntegrationParams]):
+    """Sync labels to subscription and organizations."""
+
+    @property
+    def name(self) -> str:
+        return QONTRACT_INTEGRATION
+
+    def get_clusters(self, query_func: Callable) -> list[ClusterV1]:
+        data = cluster_query(query_func)
+        return [
+            c
+            for c in data.clusters or []
+            if c.ocm is not None and integration_is_enabled(self.name, c)
+        ]
+
+    def get_organizations(
+        self, query_func: Callable
+    ) -> list[OpenShiftClusterManagerV1]:
+        return organization_query(query_func).organizations or []
+
+    def get_environments(self, query_func: Callable) -> list[OCMEnvironment]:
+        return ocm_environment_query(query_func).environments
+
+    def init_ocm_apis(
+        self,
+        environments: Iterable[OCMEnvironment],
+        init_ocm_base_client: Callable[
+            [OCMAPIClientConfigurationProtocol, SecretReaderBase], OCMBaseClient
+        ] = init_ocm_base_client,
+    ) -> dict[str, OCMBaseClient]:
+        return {
+            env.name: init_ocm_base_client(env, self.secret_reader)
+            for env in environments
+        }
+
+    def get_early_exit_desired_state(self) -> Optional[dict[str, Any]]:
+        # gqlapi = gql.get_api()
+        # return {"clusters": [c.dict() for c in self.get_clusters(gqlapi.query)]}
```

Comment:
Please remove dead code

Reply:
done

Reply:
i did not remove it but incorporate the proper desired state from all label sources